components/telemetry-operator/internal/webhook/dryrun/dryrun.go

```diff
@@ -2,122 +2,144 @@ package dryrun
+	cleanFiles, err := d.fileWriter.prepareParserDryRun(ctx, workDir, parser)
+	if err != nil {
+		return err
+	}
+	defer cleanFiles()
 
-func (v DryRunner) RunParser(ctx context.Context, configsDirectory string) error {
-	path := filepath.Join(configsDirectory, "dynamic-parsers", "parsers.conf")
-	args := append(fluentBitArgs(), "--parser", path)
-	return v.run(ctx, args)
+	path := filepath.Join(workDir, "dynamic-parsers", "parsers.conf")
+	args := append(dryRunArgs(), "--parser", path)
+	return d.runCmd(ctx, args)
 }
 
-func (v *DryRunner) run(ctx context.Context, args []string) error {
-	plugins, err := listPlugins(v.pluginDirectory)
+func (d *DryRunner) DryRunPipeline(ctx context.Context, pipeline *telemetryv1alpha1.LogPipeline) error {
+	workDir := d.newWorkDirPath()
+	cleanFiles, err := d.fileWriter.preparePipelineDryRun(ctx, workDir, pipeline)
 	if err != nil {
 		return err
 	}
-	for _, plugin := range plugins {
-		args = append(args, "-e", plugin)
-	}
+	defer cleanFiles()
 
-	out, err := runCmd(ctx, v.fluentBitPath, args...)
+	path := filepath.Join(workDir, "fluent-bit.conf")
+	args := dryRunArgs()
+	args = append(args, "--config", path)
+	externalPluginArgs, err := d.externalPluginArgs()
 	if err != nil {
-		if strings.Contains(out, "error") || strings.Contains(out, "Error") {
-			return errors.New(errDescription + extractError(out))
-		}
-		return fmt.Errorf("error while validating Fluent Bit config: %v", err.Error())
+		return err
 	}
-	return nil
+	args = append(args, externalPluginArgs...)
+
+	return d.runCmd(ctx, args)
+}
+
+func (d *DryRunner) newWorkDirPath() string {
+	return "/tmp/dry-run-" + uuid.New().String()
 }
 
-func listPlugins(pluginPath string) ([]string, error) {
+func (d *DryRunner) externalPluginArgs() ([]string, error) {
+	if d.config.FluentBitPluginDir == "" {
+		return nil, nil
+	}
+
 	var plugins []string
-	files, err := ioutil.ReadDir(pluginPath)
+	files, err := os.ReadDir(d.config.FluentBitPluginDir)
 	if err != nil {
 		return nil, err
 	}
+
 	for _, f := range files {
 		if f.IsDir() {
 			continue
 		}
-		plugins = append(plugins, filepath.Join(pluginPath, f.Name()))
+		plugins = append(plugins, filepath.Join(d.config.FluentBitPluginDir, f.Name()))
 	}
-	return plugins, err
-}
 
-func runCmd(ctx context.Context, name string, args ...string) (string, error) {
-	cmd := exec.CommandContext(ctx, name, args...)
+	var args []string
+	for _, plugin := range plugins {
+		args = append(args, "-e", plugin)
+	}
+	return args, nil
+}
 
-	outBytes, err := cmd.CombinedOutput()
+func (d *DryRunner) runCmd(ctx context.Context, args []string) error {
+	outBytes, err := d.commandRunner.run(ctx, d.config.FluentBitBinPath, args...)
 	out := string(outBytes)
-	return out, err
+	if err != nil {
+		if strings.Contains(out, "error") || strings.Contains(out, "Error") {
+			return fmt.Errorf("failed to validate the supplied configuration: %s", extractError(out))
+		}
+		return fmt.Errorf("failed to execute dryrun: %v", err.Error())
```

Comment:
This error message is shown to the user right, so I don't know if mentioning `dryrun` is the best idea since we don't describe how we validate the config

Reply:
Give me something more creative, I'll use that :) Just wanted to convey that the error is not because of failed validation or errors from the binary, but rather the command execution itself is failing.

Reply:
maybe `could not validate the supplied configuration: %v`

Reply:
@lindnerby Remember that constant we found weird? Maybe we should revert it back 🤣

Reply:
I just tested it. So if dryrun returns an error, it will print the first error: `failed to validate the supplied configuration: %s`. The second error is only printed if the fluent-bit binary returns a non-zero code and no error in the console output. It should not really happen, but there might be a bug in fluent-bit itself. I'd say, in this case we should not print the low-level error at all.

Reply:
@dennis-ge @lindnerby Please check. Also changed the wording to make it consistent with the rest of webhook errors.